amy/autoemails/actions.py

```diff
@@ -1235,3 +1245,283 @@ def get_merge_data(self) -> Optional[Dict]:
+            return None
+
+    def all_recipients(self) -> str:
+        """If available, return string of all recipients."""
+        recipients = self.recipients()
+        if recipients is not None:
+            return ", ".join(recipients)
+        return ""
+
+    def reply_to(self) -> Optional[Tuple[str]]:
+        """Overwrite in order to set own reply-to from descending Action."""
+        return (settings.ADMIN_NOTIFICATION_CRITERIA_DEFAULT,)
+
+
+class BaseProfileArchivalWarningRepeatedAction(BaseRepeatedAction):
+    """
+    Base class for emails related to profile archival.
+    """
+
+    launch_at: Optional[timedelta] = None
+    REMINDER_LIMIT: int = 3
+    EMAIL_ACTION_CLASS: Optional[Type[BaseAction]] = None
+    ACTION_NAME: Optional[str] = None
+
+    def __call__(self, *args, **kwargs):
+        people = self.get_people()
+        triggers = self.get_triggers()
+        email_reminders = self.get_email_reminders()
+        self.handle_existing_email_reminders(triggers, email_reminders, people)
+        self.create_new_email_reminders(triggers, email_reminders, people)
+
+    def get_people(self) -> Iterable[Person]:
+        """
+        Retrieves the affected users from the database.
+        """
+        return []
+
+    def get_triggers(self) -> Iterable[Trigger]:
+        """
+        Retrieves the related triggers from the database.
+        """
+        return Trigger.objects.filter(
+            active=True,
+            action=self.ACTION_NAME,
+        )
+
+    def get_email_reminders(self) -> Iterable[Trigger]:
+        """
+        Retrieves the related email reminders from the database.
+        """
+        return EmailReminder.objects.filter(
+            archived_at=None,
+            trigger__action=self.ACTION_NAME,
+        ).select_related("person")
+
+    def handle_existing_email_reminders(
+        self,
+        triggers: Iterable[Trigger],
+        email_reminders: Iterable[EmailReminder],
+        people: Iterable[Person],
+    ) -> None:
+        """
+        Update existing email reminders doing the following:
+        - Archive the email reminder where the users have logged in.
+        - Send another reminder if the user has not yet hit the REMINDER_LIMIT
+        - If the user hasn't logged in and they have hit the REMINDER_LIMIT,
+          archive the user and the email reminder.
+        """
+        people_ids = set([person.id for person in people])
+        reminders_to_archive = []
+        for reminder in email_reminders:
+            if reminder.person_id not in people_ids:
+                # if there is an email reminder that is not in people above
+                # archive the email reminder they have logged in
+                reminders_to_archive.append(reminder)
+            elif reminder.number_times_sent == self.REMINDER_LIMIT:
+                # Hit the reminder limit
+                # Archive the person and archive the reminder
+                reminder.person.archive()
```

Comment:
@pbanaszkiewicz  I'm not quite following what's being said. Would you mind writing some pseudocode so I have a good sense of what you mean? From what is said here I think that you mean:

`handle_existing_email_reminders` returns a Tuple with three lists inside `([people_to_be_archived], [reminders_to_be_archived], [reminders_to_be_sent])`

There's then a three following methods:
```
def archive_people(people_to_be_archived):
   person.archive()
def archive_reminders(reminders_to_be_archived):
   reminders.archive()
def send_reminders(reminders_to_be_sent):
    reminders.send()
```

Reply:
I'm not sure that gives cleaner code than what's written, but I may have interpreted the message incorrectly.

Reply:
Oh second thought I could see how it would be a bit easier to test against the output since you could check against the list returned. but i'd love to first know if i understood your comment correctly.

Reply:
@lauryndbrown I'll get back to you on this tomorrow.

Reply:
@lauryndbrown this is what I was thinking about:

```python
    def handle_existing_email_reminders(
        self,
        triggers: Iterable[Trigger],
        email_reminders: Iterable[EmailReminder],
        people: Iterable[Person],
    ) -> None:
        """
        Update existing email reminders doing the following:
        - Archive the email reminder where the users have logged in.
        - Send another reminder if the user has not yet hit the REMINDER_LIMIT
        - If the user hasn't logged in and they have hit the REMINDER_LIMIT,
          archive the user and the email reminder.
        """
        people_ids = set(person.id for person in people)
        reminders = set(email_reminders)
        reminders_overdue = set(reminder for reminder in reminders if reminder.number_times_sent >= self.REMINDER_LIMIT)
        reminders_confirmed = set(reminder for reminder in email_reminders if reminder.person_id not in people_ids)
        reminders_to_send = reminders - reminders_overdue - reminders_confirmed

        self.archive_profiles(reminder.person for reminder in reminders_overdue)
        self.archive_reminders(reminders_overdue | reminders_confirmed)
        self.send_reminders(reminders_to_send, triggers)

    def archive_profiles(self, people: Iterable[Person]) -> None:
        for person in people:
            person.archive()

    def archive_reminders(self, reminders: Iterable[EmailReminder]) -> None:
        EmailReminder.objects.filter(
            id__in=[r.id for r in reminders]
        ).update(archived_at=timezone.now())

    def send_reminders(self, reminders: Iterable[EmailReminder], triggers: Iterable[Trigger]) -> None:
        for reminder in reminders:
            self._update_email_reminder(reminder)
            self.schedule_rqjob(reminder, triggers)
            reminder.save()
```